macros/contexts/contextBoolean.pl

```diff
@@ -0,0 +1,160 @@
+		' *'  => { class => "Parser::BOP::and", string => ' and ', TeX => '\wedge ', perl => '&&' },
+		'* '  => { class => "Parser::BOP::and", string => ' and ', TeX => '\wedge ', perl => '&&' },
+		# implicit multiplication must have string '*' for Parser::Op (L#314)
+		' '   => { class => "Parser::BOP::and", string => ' *',    TeX => '\wedge ' },
+		'u-'  => { class => "Parser::UOP::not", string => '-',     TeX => '\mathord{\sim}' },
+		'><'  => { class => "Parser::BOP::xor", string => ' xor ', TeX => '\oplus ' },
+		'xor' => { class => "Parser::BOP::xor", string => ' xor ', TeX => '\oplus ' },
+	);
+
+	## Set default variables 'p' and 'q'
+	$context->variables->are(
+		p => 'Real',
+		q => 'Real',
+	);
+}
+
+## Subclass Value::Formula for boolean formulas
+package context::Boolean::Formula;
+our @ISA = ('Value::Formula');
+
+sub new {
+	my $class = shift;
+	my $self  = $class->SUPER::new(@_);
+
+	my @variables = $self->{context}->variables->names;
+	my @points;
+	my @values;
+
+	my $f = $self->{f} = $self->perlFunction(undef, \@variables);
+
+	foreach my $combination (0 .. 2**@variables - 1) {
+		my @point = map { $combination & 2**$_ ? 1 : 0 } (0 .. $#variables);
+		my $value = &$f(@point) ? 1 : 0;
+		push @points, \@point;
+		push @values, $value;
+	}
+
+	$self->{test_points} = \@points;
+	$self->{test_values} = \@values;
+
+	return $self;
+}
+
+package Parser::BOP::or;
+our @ISA = qw(Parser::BOP);
+
+sub _check {
+	my $self = shift;
+	# Add logic to check compatibility for logical 'or'
+	return 1;
+}
+
+sub _eval { return $_[1] || $_[2] ? 1 : 0 }
+
+sub _reduce {
+	my $self = shift;
+	# Implement reduction for logical 'or'
+	return 1;
+}
+
+sub perl {
+	my $self = shift;
+	my $perl = $self->SUPER::perl(@_);
+	return "($perl ? 1 : 0)";
+}
+
+package Parser::BOP::and;
+our @ISA = qw(Parser::BOP);
+
+sub _check {
+	my $self = shift;
+	# Add logic to check compatibility for logical 'and'
+	return 1;
+}
+
+sub _eval { return $_[1] && $_[2] ? 1 : 0 }
+
+sub _reduce {
+	my $self = shift;
+	# Implement reduction for logical 'and'
+	return 1;
+}
+
+sub perl {
+	my $self = shift;
+	my $perl = $self->SUPER::perl(@_);
+	return "($perl ? 1 : 0)";
+}
+
+package Parser::UOP::not;
+our @ISA = qw(Parser::UOP);
+
+sub _check {
+	my $self = shift;
+	# Add logic to check compatibility for logical 'not'
+}
+
+sub _eval { !($_[1]) ? 1 : 0 }
+
+sub _reduce {
+	my $self = shift;
+	# Implement reduction for logical 'not'
+}
+
+sub perl {
+	my $self = shift;
+	my $perl = $self->SUPER::perl(@_);
+	return "($perl ? 1 : 0)";
+}
+
+package Parser::BOP::xor;
+our @ISA = qw(Parser::BOP);
+
+sub _check {
+	my $self = shift;
+	# Add logic to check compatibility for logical 'not'
+}
+
+sub _eval { ($_[1] ? !$_[2] : !!$_[2]) ? 1 : 0 }
+
+sub perl {
+	my $self = shift;
+	return
+		'('
+		. '('
+		. $self->{lop}->perl(1) . ' ? ' . '!'
+		. $self->{rop}->perl(1) . ' : ' . '!!'
+		. $self->{rop}->perl(1)
+		. ') ? 1 : 0)';
+}
```

Comment:
You should be able to use the same `perl` method as you do above, if you add `perl => '=='` to the `xor` operator definitions above.

In fact, you could probably make `context::Boolean::BOP` with `@ISA = ('Parser::BOP')` and that contains the common methods (`perl`, `_check`, and `_reduce`), and then make the `@ISA` variable use `context::Boolean:BOP` rather than `Parser::BOP` in your and, or, and xor packages.  Then these would only need to have the `_eval` methods in those.

Reply:
I was actually just working on this! 👍

Reply:
That is to say, creating a parent Boolean::BOP class with the common methods is where I'm currently at. However, isn't adding `==` an issue here since we want to reverse the results? So `!=` (and we would **definitely** need _check to ensure incoming arguments are boolean in such case).